src/Interpreters/GraceHashJoin.cpp

```diff
@@ -0,0 +1,644 @@
+        {
+            if (blocks[i].rows() == 0 || i == bucket_idx)
+                continue;
+
+            if (i < bucket_idx)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected bucket index {} when current bucket is {}", i, bucket_idx);
+            buckets[i]->addRightBlock(blocks[i]);
+        }
+    }
+
+    hash_join = makeInMemoryJoin();
+
+    for (bucket_idx = bucket_idx + 1; bucket_idx < buckets.size(); ++bucket_idx)
+    {
+        current_bucket = buckets[bucket_idx].get();
+        if (current_bucket->finished() || current_bucket->empty())
+        {
+            LOG_TRACE(log, "Skipping {} {} bucket {}",
+                current_bucket->finished() ? "finished" : "",
+                current_bucket->empty() ? "empty" : "",
+                bucket_idx);
+            continue;
+        }
+
+        auto right_reader = current_bucket->startJoining();
+        size_t num_rows = 0; /// count rows that were written and rehashed
+        while (Block block = right_reader.read())
+        {
+            num_rows += block.rows();
+            addJoinedBlockImpl(std::move(block));
+        }
+
+        LOG_TRACE(log, "Loaded bucket {} with {}(/{}) rows",
+            bucket_idx, hash_join->getTotalRowCount(), num_rows);
+
+        return std::make_unique<DelayedBlocks>(current_bucket->idx, buckets, hash_join, left_key_names, right_key_names);
+    }
+
+    LOG_TRACE(log, "Finished loading all buckets");
+
+    current_bucket = nullptr;
+    return nullptr;
+}
+
+GraceHashJoin::InMemoryJoinPtr GraceHashJoin::makeInMemoryJoin()
+{
+    return std::make_unique<InMemoryJoin>(table_join, right_sample_block, any_take_last_row);
+}
+
+void GraceHashJoin::addJoinedBlockImpl(Block block)
+{
+    Buckets buckets_snapshot = getCurrentBuckets();
+    Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, block, buckets_snapshot.size());
+    size_t bucket_index = current_bucket->idx;
+
+    // Add block to the in-memory join
+    if (blocks[bucket_index].rows() > 0)
+    {
+        std::lock_guard lock(hash_join_mutex);
+
+        hash_join->addJoinedBlock(blocks[bucket_index], /* check_limits = */ false);
+        bool overflow = !fitsInMemory();
+
+        if (overflow)
+        {
+            auto right_blocks = hash_join->releaseJoinedBlocks();
+            right_blocks.pop_back();
+
+            for (const auto & right_block : right_blocks)
+                blocks.push_back(right_block);
+        }
+
+        while (overflow)
+        {
+            buckets_snapshot = rehashBuckets(buckets_snapshot.size() * 2);
+
+            blocks = JoinCommon::scatterBlockByHash(right_key_names, blocks, buckets_snapshot.size());
+            hash_join = makeInMemoryJoin();
+            hash_join->addJoinedBlock(blocks[bucket_index], /* check_limits = */ false);
+            overflow = !fitsInMemory();
+        }
+        blocks[bucket_index].clear();
+    }
+
+    if (blocks.empty())
```

Comment:
I don't see how blocks can be empty because `scatterBlockByHash()` returns number of blocks equal to number of buckets